arkouda/numeric.py

```diff
@@ -2001,3 +2027,76 @@ def value_counts(
     (array([0, 2, 4]), array([3, 2, 1]))
     """
     return GroupBy(pda).count()
+
+
+@typechecked
+def clip(
+    pda: pdarray,
+    lo: Union[numeric_scalars, pdarray],
+    hi: Union[numeric_scalars, pdarray],
+) -> pdarray:
+    """
+    Mimic the behavior of numpy.clip.
+
+    Parameters
+    ----------
+    pda : pdarray, the values to clip
+    lo  : a scalar or pdarray of the lower value(s) to clip to
+    hi  : a scalar or pdarray of the higher value(s) to clip to
+    NOTE: either lo or hi may be None, but not both.
+
+    Returns
+    -------
+    a copy of the input, where each element x remains x if lo <= x <= hi,
+                                           or equals lo if x < lo
+                                           or equals hi if x > hi
+    Notes
+    -----
+    if lo > hi, x = hi.  This is how numpy.clip behaves.
+    if all inputs are int, output is int, but if any input is real, output is real.
+      This is also how numpy.clip behaves.
+
+    Raises
+    ------
+    ValueError
+        Raised if both lo and hi are None
+    """
+
+    # see if lo or hi are None
+
+    clip_min = True if isinstance(lo, pdarray) else True if lo is not None else False
+    clip_max = True if isinstance(hi, pdarray) else True if hi is not None else False
+    if not (clip_min or clip_max):
+        raise ValueError("Either min or max must be supplied.")
+
+    # if any of the inputs are float, then make everything float
+    # note that some type checking is needed, because scalars and pdarrays get cast differently
+
+    dataFloat = pda.dtype == float
+    minFloat = (
+        True
+        if isinstance(lo, float)
+        else (True if (isinstance(lo, pdarray) and lo.dtype == float) else False)
+    )
+    maxFloat = (
+        True
+        if isinstance(hi, float)
+        else (True if (isinstance(hi, pdarray) and hi.dtype == float) else False)
+    )
```

Comment:
as a general rule of thumb, if you're using a condition to return a boolean, there's probably a way to just return the condition. I think something like this works here
```python
minFloat = isinstance(lo, float) or (isinstance(lo, pdarray) and lo.dtype == float)
maxFloat = isinstance(hi, float) or (isinstance(hi, pdarray) and hi.dtype == float)
```